### PR TITLE
Changes on Key Vault Management Swagger for Private Endpoints

### PR DESCRIPTION
**Pull request by msfcolombo**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by msfcolombo** [comment]
@tjprescott @schaabs, please let me know if some action from my side is required. Thanks

**Comment by tjprescott** [comment]
@Azure/arm-api-review-board  please take a look

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: >responses [](start = 9, length = 9)

You should specify the `x-ms-long-running-operation` since this is a long running operation ([https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation))

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: >responses [](start = 9, length = 9)

You should specify the `x-ms-long-running-operation` extension since this can be a long running operation ([[https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation)](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation))

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: >200 [](start = 11, length = 3)

Aren't you missing a 201 response status code? Creation should be 201 in general and one of the accepted patterns for long running operations uses 201+provisioningState.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: >/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/privateEndpointConnections [](start = 5, length = 141)

There doesn't appear to be a LIST operation for `privateEndpointConnections`. How would a user discover what they to a GET on in the first place?

**Review by majastrz — CHANGES_REQUESTED** [review]
I added some comments. Please take a look.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: This was copied from https://github.com/Azure/azure-rest-api-specs/blob/master/specification/storage/resource-manager/Microsoft.Storage/stable/2019-06-01/storage.json. They also don't have LIST.

Why the storage code was accepted? Is this a new requirement?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: Same as before, copied from https://github.com/Azure/azure-rest-api-specs/blob/master/specification/storage/resource-manager/Microsoft.Storage/stable/2019-06-01/storage.json. They also don't have `x-ms-long-running-operation`.

Why the storage code was accepted? Is this a new requirement?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: Same thing. The `storage.json` does not specify `201`.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (KrisBash) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: This should have a LIST operation. Nested types require both Get By Name and LIST. The fact that this was missed in another PR doesn't change that. We can work with SRP to fix.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (KrisBash) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: For all API responses in this spec, please add a "default" response to capture error cases. It should have a schema consistent with the [common error contract](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content). 
Here is an example:
https://github.com/Azure/azure-rest-api-specs/blob/0cf70260108cb40bc68d408674c187ed065f8c64/specification/batch/resource-manager/Microsoft.Batch/stable/2019-08-01/BatchManagement.json#L105

If you look at the PR for the SRP swagger you are referring to (https://github.com/Azure/azure-rest-api-specs/pull/7440#pullrequestreview-302712168) you can see that this "default" response was discussed and mandated for new APIs going forward. Please add them here.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (KrisBash) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: What is critical here is that the swagger matches the server behavior. The way this is described, the operation is always synchronous. If that is correct, no change is needed. If it is async, then a 201 response and x-ms-long-running-operation is the preferred async pattern.

**Review by KrisBash — CHANGES_REQUESTED** [review]
Left a couple comments. I think we just need the LIST api and "default" responses to sign off.  Please confirm that the PUT is synchronous.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: @KrisBash @majastrz 
Per Azure Network team definition, the LIST operation is omitted because RPs (including Storage) must include the list of `privateEndpointConnections` in the properties of the master resource. SRP does that in the definition of `StorageAccountProperties`: https://github.com/Azure/azure-rest-api-specs/blob/68831c6eded64f98fabe9eb658712ceff16c6a2d/specification/storage/resource-manager/Microsoft.Storage/stable/2019-06-01/storage.json#L1996

This PR is doing the same; see the definition of `VaultProperties`.

AKV is not adding LIST now as it does not seem necessary and would never be used by any client. If you have concerns on the design of always including child items in the properties of master resource, please open a thread with SRP, AKV and Azure Network, but don't block this PR.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: @majastrz 

Both SRP and AKV are following the guidance from https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources:

> The API flow for PUT/PATCH should be to:
> 
> 1. Respond to the initial PUT request with a 201 Created or 200 OK (per normal guidance);
> 2. (...)
> 3. **Optional:** The response headers may include a Azure-AsyncOperation header pointing to an Operation resource

I was willing to add the `x-ms-long-running-operation`, however it requires us to return 202 instead of 200 and this is a breaking change. Please do not block the PR on this.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: @majastrz  The PUT here is never used for Creation. The server never returns 201. BTW, I do agree that PUT was a poor choice of method. It should be POST, but when this was raised, other teams were already using PUT.

@KrisBash Regarding `x-ms-long-running-operation`, please see https://github.com/Azure/azure-rest-api-specs/pull/7631#discussion_r363998777.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: Done. This was doable because the service does return 202.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: Done.

**Review by schaabs — COMMENTED** [review]
Shouldn't these changes be added to the 2019-09-01 specification, or possibly a new preview API version?

**Comment by msfcolombo** [comment]
@schaabs My fork doesn't have the `2019-09-01`. Can you specify which fork are you looking at?

These are all incremental changes and all RPs are supposed to provide near identical swagger changes, meaning there is little room for modification. Given that, there is no need for preview version.

**Comment by schaabs** [comment]
@msfcolombo `2019-09-01` is in the master branch of the repo, you'll need to pull it from the azure/master

**Comment by msfcolombo** [comment]
@schaabs @KrisBash @majastrz 

I've answered all your inquiries with comments or modifications. Please take another look.

Thanks

**Comment by tjprescott** [comment]
@Azure/arm-api-review-board please take a look.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (sanjaiganesh) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: PUT is required for template deployments.. POST is not idempotent. So, template deployments use only PUT. 

---
In reply to: [363999344](https://github.com/Azure/azure-rest-api-specs/pull/7631#discussion_r363999344) [](ancestors = 363999344)

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (sanjaiganesh) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: > "x-ms-long-running-operation": true [](start = 7, length = 36)

also add following to help SDK know which one to use for tracking LRO

add "x-ms-long-running-operation-options" extension with "final-state-via" property.
``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "location"
}
```

or

``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "azure-async-operation"
}
```

Comment (sanjaiganesh) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: same with other LROs (returning 202 or 201)

---
In reply to: [368137466](https://github.com/Azure/azure-rest-api-specs/pull/7631#discussion_r368137466) [](ancestors = 368137466)

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (sanjaiganesh) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: <while I am not following full thread here>
The feedback is per Azure guidelines and for better customer experience (to not let them force different ways of doing things for different RP / types). I am sure we all agree that is important, instead of following what SRP did in past.

---
In reply to: [363992368](https://github.com/Azure/azure-rest-api-specs/pull/7631#discussion_r363992368) [](ancestors = 363992368)

**Comment by tjprescott** [comment]
@sanjaiganesh since you have signed off for ARM, can you dismiss reviews for @KrisBash and @majastrz  (assuming their comments are addressed). I cannot dismiss their review, nor can I merge the PR with any "changes requested".

**Review by tjprescott — APPROVED** [review]

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: I hear you. As explained, I don't own API design of Private Links. I just own the product that needs to integrate. Some of these APIs are be called by NRP, so we have little room to change.

IMHO, the problems appeared when changes to SRP were approved in Swagger. They are now in `stable` state, NRP is deployed globally, and I don't think this would ever change in near future.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json
Comment (msfcolombo) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2018-02-14/keyvault.json]: Yes. This API is not used for template deployment.

**Comment by ShaneBala-keyvault** [comment]
@sanjaiganesh ping on this, since ARM has signed off, can we dismiss the remaining issues and merge this PR?

**Review by tjprescott — APPROVED** [review]

**Comment by tjprescott** [comment]
Yes, I can re-approve the PR, but until those reviews are accepted or dismissed (I can't do either) I can't merge this.

**Comment by yungezz** [comment]
hi @NullMDR , could you pls help to look at why the Ignore task blocking PR merging? thanks.

**Comment by NullMDR** [comment]
It's OK. Good to merge.

**Pull request merged**
